### PR TITLE
Allow for externally-supplied `storage_options` to S3 loader and further down to Reductionist

### LINKED ISSUE #177: Create storage_options and active_storage_url API
Create storage_options and active_storage_url API so that they are not  hardwired

### PR DESCRIPTION
**Pull request by valeriupredoi**
## Description

This is a locally-constructed, and off the back of #180 version of #178 - merely translating @davidhassell 's work here, but with a twist :grin: 

What we did here:

- `Active` now ingests two new kwargs `storage_options=None (type: dict)` and `active_storage_url=None (type: str)` by means of which a user can specify S3 credentials straight into Active (or pass them from a higher level eg `cf-python`). These can be varied, and tested for - e.g. a collection of items like the credentials needed to get into a standard S3 bucket, or an anon=True bucket. See tests/test_compression.py and tests/test_compression_remote_reductionist.py - there we use `storage_options` and `active_storage_url` kwargs to perfectly reproduce connecting to a localhost/Minio data object from a local Reductionist, and also to connect to a valid yet anon=True bucket via a remote Reductionist (this latter stops at Access Denied since Reductionist is not yet able to support anon=True buckets)
- these storage options have now been integrated in the open mechanisms in `Active` and `netcdf_to_zarr`

Closes #177  and the valiant #178 

## Before you get started

- [x] [☝ Create an issue](https://github.com/valeriupredoi/PyActiveStorage/issues) to discuss what you are going to do

## Checklist

- [x] This pull request has a descriptive title and labels
- [x] This pull request has a minimal description (most was discussed in the issue, but a two-liner description is still desirable)
- [x] Unit tests have been added (if codecov test fails)
- [x] Any changed dependencies have been added or removed correctly (if need be)
- [x] All tests pass

### CODE REVIEW

File: activestorage/active.py [hunk 1]
```diff
@@ -406,22 +448,39 @@ def _process_chunk(self, session, fsref, chunk_coords, chunk_selection, counts,
...
+                                                       S3_ACTIVE_STORAGE_URL,
+                                                       S3_URL,
+                                                       bucket, object, offset,
+                                                       size, compressor, filters,
+                                                       missing, dtype,
+                                                       self.zds._chunks,
+                                                       self.zds._order,
+                                                       chunk_selection,
+                                                       operation=self._method)
+            else:
```
Comment (davidhassell) [inline activestorage/active.py:470]: Why do we need a special case for `if not self.storage_options:`? See previous comment on pre-configured options ...

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi V - looks fine to me, apart from the whole pre-configured options thing. I'm in the office tomorrow if you want to chat about it.

File: activestorage/active.py (hunk 1 above)
Comment (valeriupredoi) [reply activestorage/active.py:470]: hey bud @davidhassell - also RE your comment above: very good question and counter-intuitive at first glance indeed, I agree. But here's my logic behind it:
- let's keep a hardwired configuration for now, test the StackHPC MinIO implemented functionality continually - so far, it's the only S3 configuration that actually works
- when we have tested with the `storage_options` and are sure of its functionality and generality, my intention is still to keep that configuration that we have now, but make it available through not reading an `activestorage/config.py` module but rather a number of set piece configuration yaml files that we have pre-configured for various sites - organizations can contribute those once they have S3 object stores

Let's talk about it tomorrow, shall we?

**Comment by valeriupredoi** [comment]
OK we are in very good shape - this thing does a dance and half, pretty much toasts the bread and spreads the butter too, don't mind the S3 tests failing, **those are supposed to fail** given Reductionist doesn't support `anon=True` buckets, but the cf-python integration and all else is there. Needs a bit more unit testing, but let's first talk about it tomorrow :beer:

File: activestorage/active.py [hunk 2]
```diff
@@ -34,10 +34,15 @@ def load_from_s3(uri):
...
     """
-    fs = s3fs.S3FileSystem(key=S3_ACCESS_KEY,  # eg "minioadmin" for Minio
-                           secret=S3_SECRET_KEY,  # eg "minioadmin" for Minio
-                           client_kwargs={'endpoint_url': S3_URL})  # eg "http://localhost:9000" for Minio
+    if not storage_options:  # use pre-configured S3 credentials
+        fs = s3fs.S3FileSystem(key=S3_ACCESS_KEY,  # eg "minioadmin" for Minio
+                               secret=S3_SECRET_KEY,  # eg "minioadmin" for Minio
+                               client_kwargs={'endpoint_url': S3_URL})  # eg "http://localhost:9000" for Minio
+    else:
+        fs = s3fs.S3FileSystem(**storage_options)  # use passed-in dictionary
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply activestorage/active.py:45]: V and I chatted about this, and I'm fine leaving this for now. We'll review prior to the public release.

File: activestorage/active.py [hunk 3]
```diff
@@ -406,22 +449,44 @@ def _process_chunk(self, session, fsref, chunk_coords, chunk_selection, counts,
...
-            tmp, count = reductionist.reduce_chunk(session, S3_ACTIVE_STORAGE_URL,
-                                                   S3_URL,
-                                                   bucket, object, offset,
-                                                   size, compressor, filters,
-                                                   missing, dtype,
-                                                   self.zds._chunks,
-                                                   self.zds._order,
-                                                   chunk_selection,
-                                                   operation=self._method)
+            if not self.storage_options:
```
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply activestorage/active.py]: done in https://github.com/valeriupredoi/PyActiveStorage/pull/182/commits/c37694ecba67352927455c03f1311406fcb1cd23 and https://github.com/valeriupredoi/PyActiveStorage/pull/182/commits/56547ce97e341b10e1d7e8b20d0dbc28bdc0beb0 :beer:

**Review by davidhassell — APPROVED** [review]
Done

**Comment by valeriupredoi** [comment]
> Done

mega, many thanks @davidhassell :beers:

**Pull request merged**